### ISSUE TITLE
build gutenberg block for the default form

### ISSUE THREAD

**Issue by selul** [Codeinwp/pirate-forms#226]
I would like to test the ground by building a block for the pirate_forms, the default form. 
 
We can add this new block  in Blocks sections -> Widgets. 

For this first iteration, the block wont have any settings, it will just render the pirate_forms default form. 

If you have any problems in building this dynamic block, you can check the similar ones ( Latest posts, Categories) 

You can work on this branch **gt-integration**  branch

**Comment by selul** [comment]
@contactashish13 let's put this on hold for the moment.

**Comment by HardeepAsrani** [comment]
@selul Would love to work on this one, if you want. :)

**Comment by selul** [comment]
@HardeepAsrani we have some WIP done by @andreilupu on a different project that might be reused here. Let's see how that is going before starting from scratch here. 

Also, if you are interested in practice Gutenberg we can start the work for wppr, feedzy or visualizer, one of your choice. What do you think ?

**Comment by HardeepAsrani** [comment]
@selul Sure, would love that.

**Comment by selul** [comment]
@HardeepAsrani let's choose one that you are most comfortable with.

**Comment by HardeepAsrani** [comment]
I'll try with Feedzy first. I'll be off this following week, to Thailand, and will work as soon as I get back. Will get all of them ready by the end of March.

**Comment by selul** [comment]
@HardeepAsrani great, we can have a call before you will get started on this.

**Comment by HardeepAsrani** [comment]
@selul Today of sometime else? After today I'll be out for a week.

**Comment by selul** [comment]
Let's do it when you will be back before you will start working on it.
Enjoy WordCamp Bangkok! :D

**Comment by rodica-andronache** [comment]
@HardeepAsrani you started this on feedzy? :D How is it going? Could we move the task there, to leave this just for Pirate Forms? 👍

**Comment by HardeepAsrani** [comment]
@rodica-andronache Feedzy is pretty much done. Gutenberg part is done, some issues with PHP part which Marius is taking care of. And this ticket can be used for Pirate Forms. :p

**Comment by rodica-andronache** [comment]
@contactashish13 how is this going?

**Comment by contactashish13** [comment]
@rodica-andronache I've added basic functionality where one can add the default form and save it so that it displays on the front end. PR: https://github.com/Codeinwp/pirate-forms/pull/300

This would need a lot of work depending on the UX we eventually want (including the behavior for PRO).

**Comment by rodica-andronache** [comment]
@contactashish13 looking good 👍 
What I think we need next is:
- have some sort of basic styling, so we can release it. Maybe @HardeepAsrani can help, if needed 👍 
- add a new block for the PRO version with a select of forms created with the pro version.
What do you think?

**Comment by HardeepAsrani** [comment]
@contactashish13 You used an amazing hack to display form's HTML without much hassle. Would have never imagined that in a billion years. 😛 

- For the styles, we can just enqueue `front.css` file to Gutenberg, which will appear like this:

![screenshot-block local-2018-05-22-03-26-57](https://user-images.githubusercontent.com/2649903/40332052-3c1462ae-5d70-11e8-8549-8d60b7a2cc1b.png)

It will look pretty much same as front-end...but without any styles for the button.

**Comment by rodica-andronache** [comment]
It looks good 👍

**Comment by contactashish13** [comment]
@rodica-andronache have fixed this as well as added support for pro. The inspector panel can be used to change the form. I think this needs extensive testing 

- with/without pro
- with/without gutenberg 
- with multiple forms

**Comment by rodica-andronache** [comment]
@contactashish13 I had just a brief look. I think there is a small problem with the Custom spam trap in the dashboard https://prnt.sc/jpnzvf. When using recaptcha, is not showing on the dashboard.
And when choosing Edit as HTML, you can't actually edit the form. I'm not sure how we can do this, but how this is working now, I'm not sure there is any way of using the ajax support in the form.  Do you have any suggestions?

**Comment by rodica-andronache** [comment]
@rohittm please have a round of testing for this, and let me know what you find out in each case 👍

**Comment by contactashish13** [comment]
@rodica-andronache good point about the ajax support! :+1: have added that in the inspector panel as well and fixed all other issues you mentioned. I have removed the Edit as HTML as I don't think it was of any use.

**Comment by rohittm** [comment]
@contactashish13 I tested with the free version. It works fine. One thing I found was:

Step 1: Enable spam trap (custom, or reCAPTCHA) in settings.
![screen shot 2018-06-02 at 18 09 19](https://user-images.githubusercontent.com/3204558/40874688-1d2601bc-6690-11e8-930e-777ad81273a5.png)
Step 2: Add a new post with pirate forms block in Gutenberg and publish it.
![screen shot 2018-06-02 at 18 10 51](https://user-images.githubusercontent.com/3204558/40874706-5289b15a-6690-11e8-900b-d80fca5e7a27.png)
Step 3: Change the spam trap settings (No/custom/reCAPTCHA), and not the one set in step 1.
![screen shot 2018-06-02 at 18 11 36](https://user-images.githubusercontent.com/3204558/40874717-694a6b6e-6690-11e8-8b6e-546d53ded582.png)
Step 4: Edit the same post made in step 2. 
![screen shot 2018-06-02 at 18 12 00](https://user-images.githubusercontent.com/3204558/40874720-8cb11fb2-6690-11e8-83b0-2f0a905c2087.png)

You'll see the block will continue to display old spam trap, even if it is changed in the settings. Although it works fine with the frontend, the spam trap settings aren't changed until you delete and add the block again. If I add a new block below the old one, the new block is with the new settings, but the old one stays the same. 
![screen shot 2018-06-02 at 18 08 58](https://user-images.githubusercontent.com/3204558/40874742-d36e4b1e-6690-11e8-8acc-3eb2450acb86.png)

Yes, I cleared cache, and cookies opened the same post in private window, it shows the same behavior. 

Now, when I tested PRO version of the form, everything worked fine except for the spam tram option. The above problem with the default form exists. Here are the steps:

Step 1: Add a new form, add whatever fields you like and add spam trap: Google reCAPTCHA. When you add the site and secret keys, save the form.
Step 2: Add a new post with pirate forms block and add the form (PRO one) you created.
Step 3: You'll be able to see the spam trap in with the form. Save the post.
Step 4. Open the post in the frontend, and the spam trap is missing.

[Here's](https://drive.google.com/file/d/1caxLrGv2BE_ep8GnsD2Zimdtf4swasVK/view?usp=sharing) a video to demonstrate this.

I'd also like to suggest if we can have form settings in inspector element of the block for the ease of access, and it'll help in knowing what the form will look like. What do you think?

cc @rodica-andronache

**Comment by contactashish13** [comment]
@rohittm yes this is a problem for which (in my limited knowledge) there is no solution because gutenberg saves the HTML in the editor. The same behavior will be exhibited if you change a field's display type from required/display to not display. @HardeepAsrani can you please comment here?

**Comment by HardeepAsrani** [comment]
Unless you fetch things every time the page loads, this will happen, unfortunately. Don't think it will affect anyone much. 😄

**Comment by contactashish13** [comment]
@HardeepAsrani I've changed this to fetch the form every time but it is not the best implementation because any time I change an attribute it thinks the form has changed and prompts me to save the post. Can you please check if there is a better way?

**Comment by HardeepAsrani** [comment]
Don't see any other ways. It's better to let it go as it's not a deal breaker.

**Comment by rodica-andronache** [comment]
@rohittm I would need you to test again the gutenberg block here https://github.com/Codeinwp/pirate-forms/pull/320 . Please test the Ajax option and see if everything is ok.
Not sure if it's because of gutenberg, or my computer, but every time I try to enable ajax for the block, my page gets stuck.

**Comment by rohittm** [comment]
@rodica-andronache Same thing is happening with me as well. AJAX forms aren't working. And when I go back to the editor, the page gets stuck, and I am unable to scroll.

cc @contactashish13

**Comment by contactashish13** [comment]
@rohittm can you share a video?

**Comment by HardeepAsrani** [comment]
@contactashish13 I think Rohit was talking about this as it's happening with me as well: https://drive.google.com/file/d/1vYiEeZSuoQUJ9cnlIteExDOrwFwrhbf_/view?usp=sharing

**Comment by HardeepAsrani** [comment]
I think I've figured out the issue, so will try to debug it. It's ending requests to Rest API every moment which is causing this issue.

**Comment by HardeepAsrani** [comment]
@contactashish13  Seems like a lot of functions are being called again and again. It's an endless loop. Maybe you can try add some comment to each function and see this loop. It happens when you turn from Text to Visual editor.

I'm testing with free version only. I can try to solve it but I don't understand logic used here for loading so that will be hard. :) If you can't reproduce this on your end, then we can have a call and you can explain me the logic used and then I can try to debug. It has to do something with `html_changed` being set to true again and again where you're doing `if(props.attributes.html === data.html){` check. It shouldn't be the case, but for some reason it's happening.

It will be easier to make the code as a React component so you could use React's Lifecycle methods.

Also, please commit `.babelrc` file because I spent 10 mins banging my head around on why the build was failing, something you experienced when you were using my code last time. 😆 

Apart from that, awesome job making it way less complicated than I had originally imagined. Hope my reply wasn't as confusing as it sounds to me. 😄

**Comment by contactashish13** [comment]
@rodica-andronache waiting for @HardeepAsrani's response
